### ISSUE TITLE
Pre-paring for data

### ISSUE THREAD

**Issue by John1231983** [marianocabezas/challenges2017#1]
Hello, I found that you are using iseg dataset. I found one mistake in your implementation. The number of class should be 4: CSF, WM, GM, and BG, instead of 5 in your implementation
In additions, I found that you did not apply normalization. Actually, the deep learning needs to normalize the data to zero mean and unit variance. Do you consider the step? 
Hope it help!

**Comment by marianocabezas** [comment]
Hi,
While it is true I have 5 classes defined at the start of the main function (thanks, I actually didn't notice that bug), it does not really affect the network since all the labels range from 0 to 3 anyways (the output for label 4 has always 0). I had that value due to copy+pasting part of the code and not checking it.
About normalization, I am applying it. I have a function just to do that. It normalizes each voxel according to the mean and standard deviation of the whole image. I also added batch normalization recently just to see its effect.

Thanks!

**Comment by John1231983** [comment]
You are welcome. 
For normalization, we have some options:
1. Normalize slice by slice. For each slice, we will compute mean and std and then do as `slice_i=(slice_i-mean)/std`
2. Normalize for whole image. We will compute mean and std of whole image(inclide black/zero slice) and do as `image_normalize=(image-mean)/std`
3. Normalize from all images in training set. We will compute mean and std based on all images in training set, instead of an image itself. Then we normalize each image based on these value.

Which option do you prefer? I prefer the second option. Could you test which one is best for you? Thanks

**Comment by marianocabezas** [comment]
I am already using option 2. I have been using it for a long time for any network I have tried. I don't like option 1 because it produces artifacts and about the 3rd option, I have used it on other applications, but using histogram matching.

**Comment by John1231983** [comment]
Me too. I like the second option, but for iseg dataset that includes many zero slices (all pixels are zero), so I did not think the second slice is best one. A good pre-processing step may gain 0.5%. A solution to ignore zero slices is cutting/cropping region, but it may hurt the background prediction in the testing phase( I did it). Hence, I am testing in the third option but I am not sure it is good for iseg dataset.  Do you use crop region (remove zero slices) when normalization?

**Comment by marianocabezas** [comment]
I do not take zero values into account when normalizing. That way I am sure the mean and standard deviation are ok.

**Comment by John1231983** [comment]
Thanks for your good point. But, how could you do it? For example, we have two sequences T1 and T2: `img_T1, img_T2`. Each sequence has size 144x192x256. The normalize images denote `img_T1_normalize, img_T2_normalize`. So, how to implement these normalized images in python? I am very interesting because we also need to normalize in the testing phase

**Comment by marianocabezas** [comment]
Using numpy is easy: `img_T1_normalize = (img_T1 - img_T1[np.nonzero(img_T1)].mean()) / img_T1[np.nonzero(img_T1)].std()`. If you want you can also do `img_T1_nonzero = img_T1[np.nonzero(img_T1)]` and use this variable instead of repeating the nonzero part twice (it should be mor epythonic this way).

**Comment by John1231983** [comment]
Thanks for your solution. Given an image in iSeg dataset with size of 144x192x256, it has about 20 slices which are zero slices. Using your code `img_T1_normalize = (img_T1 - img_T1[np.nonzero(img_T1)].mean()) / img_T1[np.nonzero(img_T1)].std()`, so we don't need to crop/cutting zero region (likes removing zero slices to image size of 144x192x236). Is it right?

**Comment by marianocabezas** [comment]
Yes, the `np.nonzero` function returns a mask of all the voxels that have a value different than zero.

**Comment by John1231983** [comment]
@marianocabezas : I tested several  schemes: (1) normalize without zero position, (2) normalize the whole image with zero position, (3) normalize with cropped image (removing some zero slice, using a rectangular near brain), (4) subtract mean (computing from whole subjects) and divide by maximum values (1000 with iseg). 

I found that the last option is best for me. It gains about 1%. I hope that information is useful for you. The reason what I think is iseg dataset is not so variance, the pixel intensity range of data is often from 0 to 1000. Hence, a simple pre-processing step will be good, without losing distribution. 

The last thing I want to ask about data augmentation. It often uses in limited dataset. However, the brain is synthetic, hence the augmentation methods (flip, rotation) may not be helpful (as I tested). Do you think so? What kind of augmentation method did you use for iseg dataset?

**Comment by marianocabezas** [comment]
About the schemes... Don't always trust accuracy/loss of the training dataset without checking if it stands with the validation dataset. Also, it does not always translate to better DSC/Hausdorff results (which is what the organizers evaluate).

Abou data augmentation it really depends on the architecture and your input/output shapes. For a brain image, sometimes flipping the x axis works (not always), but rotation and translation might not... There isn't an easy answer.

For the record, I rarely use data augmentation with brain images.

**Comment by John1231983** [comment]
Hello again, Long time no see. I would like to ask one question. I don't want to open new issue because it is not bug

My question related to `modified_hausdorff_distance` function at [here](https://github.com/marianocabezas/data_manipulation/blob/49d408ed1b35280f451e5ff9fd49816b9461e6da/metrics.py). Is it same meaning with `95th-percentile of the Hausdorff distance` in the paper http://www.sciencedirect.com/science/article/pii/S1053811914010349? I used your code to measure MHD but it is not same as the result as the challenge. The challenge report about more than 5., while your MHD is 0.4 . Thanks

**Comment by marianocabezas** [comment]
Hi again!

The measure included in the file is D<sub>22</sub> as expressed in the following paper: http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.1.8155&rep=rep1&type=pdf. I believe we wanted to use to compare results with a paper that used it, but I don't really remember which paper it was. However, it is not the 95th-percentile one that is being used currently in some workshops or the one from the paper you cited.
